sycl/doc/extensions/FreeFunctionQueries/FreeFunctionQueries.asciidoc

```diff
@@ -0,0 +1,199 @@
+supporting this extension must predefine the macro `SYCL_EXT_ONEAPI_FREE_FUNCTION_QUERIES`
+to one of the values defined in the table below. Applications can test for the
+existence of this macro to determine if the implementation supports this
+feature, or applications can test the macro's value to determine which of the
+extension's APIs the implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value |Description
+|1     |Initial extension version. Base features are supported.
+|===
+
+== Overview
+
+The extension allows to get `sycl::id`, `sycl::item`, `sycl::nd_item`,
+`sycl::group` and `sycl::sub_group` instances globally, without having to
+explicitly pass them as arguments to each function used on the device.
+
+NOTE: Passing such instances as arguments can result in a clearer interface
+that is less error-prone to use. For example, if a function accepts a
+`sycl::group`, the caller must assume that function may call a
+`sycl::group_barrier` and ensure that associated control flow requirements are
+satisfied. It is recommended that this extension is used only when modifying
+existing interfaces is not feasible.
+
+== Accessing Instances of Special SYCL Classes
+
+The `sycl::ext::oneapi::this_kernel` namespace contains functionality related
+to the currently executing kernel.
+
+It is the user's responsibility to ensure that that these functions are called
+in a manner that is compatible with the kernel's launch parameters, as detailed
+in the definition of each function. Calling these functions from an incompatible
+kernel results in undefined behavior.
+
+[source,c++]
+----
+namespace sycl {
+namespace ext {
+namespace oneapi {
+namespace this_kernel {
+
+size_t get_linear_id();
+
+template <int Dimensions>
+id<Dimensions> get_id();
+
+template <int Dimensions>
+item<Dimensions> get_item();
+
+template <int Dimensions>
+nd_item<Dimensions> get_nd_item();
+
+template <int Dimensions>
+group<Dimensions> get_group();
+
+sub_group get_sub_group();
+
+}
+}
+}
+}
+----
+
+[source,c++]
+----
+size_t get_linear_id();
+----
+_Returns_: A `size_t` representing the current work-item's linear position in
+the global iteration space. Multi-dimensional indices are converted into a
+linear index following the linearization equations defined in Section 3.11.1 of
+the SYCL 2020 specification.
+
+[source,c++]
+----
+template <int Dimensions>
+id<Dimensions> get_id();
+----
+_Preconditions_: `Dimensions` must match the dimensionality of the currently
+executing kernel. The currently executing kernel must have been launched with a
+`sycl::range` or `sycl::nd_range` argument.
+
+_Returns_: A `sycl::id` instance representing the current work-item in a
+`sycl::range`.
+
+[source,c++]
+----
+template <int Dimensions>
+item<Dimensions> get_item();
+----
+_Preconditions_: `Dimensions` must match the dimensionality of the currently
+executing kernel. The currently executing kernel must have been launched with a
+`sycl::range` or `sycl::nd_range` argument.
```

Comment:
That's a good question, and I don't know what's preferred here.

The SYCL specification only mentions two execution models [here](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#_sycl_kernel_execution_model), and says that "Each work-item is identified by a value of type `item<N>`". So, a user might reasonably expect all kernels to support either `item` or `nd_item`.

I can see an argument that an implementation might have more optimization freedom if it can special-case compilation of kernels taking `id` arguments... But it also seems like it might add more implementation complexity.  If we say that it's illegal to access an `item` when the kernel's parameter is an `id`, we'd effectively be forcing implementations to start distinguishing between three types of kernel (instead of only two).

Reply:
I'm not sure either.  I don't think this would require the implementation to be more complex.  It simply allows an implementation to be more complex (and gain some optimization) if it wants to.

Does DPC++ currently optimize the case when the kernel takes an `id` rather than an `item`?

Reply:
Yes, agreed that an implementation would need to be more complex if it wanted to flag an error.

I assume your conclusion, therefore, is that `get_item()` can be called even from a kernel that is declared to take an `id` parameter?

Reply:
Yes.  I think it's less cognitive burden for users (because there are only two execution models to worry about) and simpler for implementers in the long run (because introducing good error behavior later is desirable).

Do you think we need a non-normative note confirming that we're deliberately allowing `get_item()` to be called from a function accepting an `id`?

Reply:
I don't think it is necessary, but I am not opposed.

Reply:
Ok, I'll leave it as it is for now then.  If anybody is interested they can always read this discussion!